lib/providers/openstreetmap.js

```diff
@@ -0,0 +1,172 @@
+var _ = require('lodash');
+
+var vumigo = require('vumigo_v02');
+var JsonApi = vumigo.http.api.JsonApi;
+
+var provider = require('./provider');
+var Provider = provider.Provider;
+var AddressResult = provider.AddressResult;
+
+var OpenStreetMap = Provider.extend(function (self, opts) {
+    /**class:OpenStreetMap(opts)
+
+    An address search provider for Open Street Map.
+
+    :param string opts.api_url:
+        The URL of the Open Street Map API. Defaults to
+        'http://open.mapquestapi.com/nominatim/v1/search.php'.
+
+    :param function opts.extract_address_label:
+        Function that extracts a human-friendly label from an Open Street Map
+        API result. See :meth:`OpenStreetMap.extract_address_label` for the
+        function signature and default implementation.
+
+    :param function opts.extract_address_data:
+        Function that extracts data to store on a contact from an Open Street Map
+        API result. See :meth:`OpenStreetMap.extract_address_data` for the
+        function signature and default implementation.
+
+    :param boolean hard_boundary:
+        Whether to limit results to a fixed bounding box. Defaults to `true`.
+
+    :param integer address_limit:
+        Maximum number of search results to return. Defaults to `30`.
+
+    :param array bounding_box:
+        Bounding box to limit search results to. Defaults to
+        `["-180.0", "90.0", "180.0", "-90.0"]`.
+    */
+
+    // Defaults
+    opts = _.defaults(opts || {}, {
+        api_url: 'http://open.mapquestapi.com/nominatim/v1/search.php',
+        extract_address_label: null,
+        extract_address_data: null,
+        hard_boundary: true,
+        address_limit: 30,
+        bounding_box: ["-180.0", "90.0", "180.0", "-90.0"],
+    });
+
+    Provider.call(self);
+
+    // Attributes
+    self.api_url = opts.api_url;
+    self.hard_boundary = opts.hard_boundary;
+    self.address_limit = opts.address_limit;
+    self.bounding_box = opts.bounding_box;
+    self.http = null;
+
+    self.init = function(im) {
+        /**:OpenStreetMap.init()
+
+        Initialization function invoked during state initialization.
+
+        :param InteractionMachine im:
+            The state's :class:`InteractionMachine` instance.
+        */
+        self.http = new JsonApi(im);
+    };
+
+    self.extract_address_label = opts.extract_address_label || function(result) {
+        /**:OpenStreetMap.extract_address_label(result)
+
+        Returns the value of ``result.display_name`` as a human-friendly
+        display name for an Open Street Map location result.
+
+        May be overriden using the ``extract_address_label`` class option.
+
+        :param object result:
+            A location result from the Open Street Map API.
+
+        :return string:
+            The label to display for this result.
+        */
+        return result.display_name;
```

Comment:
Best I can suggest now is `"road", "suburb", ["city" || "town" || "village"], "postcode", "country"` .  It's a bit messy because the data structure isn't fixed.  Sometimes those fields will be `undefined`.